shim/src/glue.c

```diff
@@ -1,23 +1,38 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
-
+#define __GNU_SOURCE
 #include <demi/libos.h>
 #include <demi/sga.h>
 #include <demi/wait.h>
 
-static __thread int __demi_reent_guard = 0;
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <sys/types.h>
+
+#include "utils.h"
+#include "log.h"
+
+static struct hashset *__demi_reent_guards;
 
-#define DEMI_CALL(type, fn, ...)    \
-    {                               \
-        __demi_reent_guard = 1;     \
-        type ret = fn(__VA_ARGS__); \
-        __demi_reent_guard = 0;     \
-        return (ret);               \
+#define DEMI_CALL(type, fn, ...)                        \
+    {                                                   \
+        pid_t tid = getpid();                           \
```

Comment:
Well, this actually gives you the process identifier, which should be the same for all threads... Did you mean `gettid()` instead?

Reply:
I did, I changed it back to `gettid()` now.